regression serde feature does not work with no-std

With the release of v0.19 we broke the ability to use Serde in no_std contexts. When you compile with:

```console
$ cargo build --no-default-features --features serde
   Compiling multihash v0.19.0 (/tmp/rust-multihash)
error[E0432]: unresolved import `std`
 --> src/serde.rs:3:5
  |
3 | use std::fmt;
  |     ^^^ use of undeclared crate or module `std`

error: cannot find macro `format` in this scope
  --> src/serde.rs:35:46
   |
35 |             .map_err(|err| de::Error::custom(format!("Failed to deserialize Multihash: {}", err)))
   |                                              ^^^^^^

error: cannot find macro `format` in this scope
  --> src/serde.rs:48:46
   |
48 |             .map_err(|err| de::Error::custom(format!("Failed to deserialize Multihash: {}", err)))
   |                                              ^^^^^^

error[E0599]: no method named `to_bytes` found for reference `&Multihash<SIZE>` in the current scope
  --> src/serde.rs:17:42
   |
17 |         serializer.serialize_bytes(&self.to_bytes())
   |                                          ^^^^^^^^ method not found in `&Multihash<SIZE>`

error[E0433]: failed to resolve: use of undeclared type `Vec`
  --> src/serde.rs:43:25
   |
43 |         let mut bytes = Vec::new();
   |                         ^^^ use of undeclared type `Vec`

Some errors have detailed explanations: E0432, E0433, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `multihash` (lib) due to 5 previous errors
```

This is a regression as it works with v0.18.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -60,6 +60,22 @@ jobs:
         run: cargo build --no-default-features --workspace --target thumbv6m-none-eabi
         shell: bash
 
+  build-no-std-with-serde:
+    name: Build no_std with `serde` feature enabled
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Sources
+        uses: actions/checkout@v3
+
+      - name: Install Rust Toolchain
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          targets: thumbv6m-none-eabi
+
+      - name: Build
+        run: cargo build --no-default-features --features serde --workspace --target thumbv6m-none-eabi
+        shell: bash
+
   msrv:
     name: MSRV
     runs-on: ubuntu-latest

--- a/codetable/Cargo.toml
+++ b/codetable/Cargo.toml
@@ -28,7 +28,7 @@ strobe-rs = { version = "0.8.1", default-features = false, optional = true }
 ripemd = { version = "0.1.1", default-features = false, optional = true }
 multihash-derive = { version = "0.9.0", path = "../derive", default-features = false }
 core2 = { version = "0.4.0", default-features = false }
-serde = { version = "1.0.158", features = ["derive"], optional = true }
+serde = { version = "1.0.158", features = ["derive"], default-features = false, optional = true }
 
 [dev-dependencies]
 hex = "0.4.2"

--- a/src/serde.rs
+++ b/src/serde.rs
@@ -1,20 +1,76 @@
 //! Multihash Serde (de)serialization
 
-use std::fmt;
+use core::{fmt, ptr, slice};
 
 use serde::{
     de::{self, SeqAccess, Visitor},
-    Deserialize, Deserializer, Serialize, Serializer,
+    ser, Deserialize, Deserializer, Serialize, Serializer,
 };
 
 use crate::Multihash;
+
+/// The maximum serialization size of `code` is 9 bytes (a large varint encoded u64) and for `size`
+/// is 2 bytes  (a large varint encoded u8), this makes a total of 11 bytes.
+const MAXIMUM_PREFIX_SIZE: usize = 11;
+
+/// The is currently no way to allocate an array that is some constant size bigger then a given
+/// const generic. Once `generic_const_exprs` are a thing, this struct will no longer be needed.
+/// Until then we introduce a hack. We allocate a struct, which contains two independent arrays,
+/// which can be specified with const generics. We then treat the whole struct as a slice of
+/// continuous memory.
+#[repr(C, packed)]
+struct Buffer<const SIZE_FIRST: usize, const SIZE_SECOND: usize> {
+    first: [u8; SIZE_FIRST],
+    second: [u8; SIZE_SECOND],
+}
+
+#[allow(unsafe_code)]
+impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
+    fn new() -> Self {
+        let buffer = Self {
+            first: [0; SIZE_FIRST],
+            second: [0; SIZE_SECOND],
+        };
+
+        // Make sure that the struct allocated continuous memory, as we exploit that fact with the
+        // `as_slice` and `as_mut_slice()` methods.
+        let start_first = ptr::addr_of!(buffer.first) as *const u8;
+        let start_second = ptr::addr_of!(buffer.second) as *const u8;
+        unsafe {
+            // This should never happen, hence it's OK to panic.
+            if start_second.offset_from(start_first) != SIZE_FIRST as isize {
+                panic!("alignment of the struct is wrong")
+            }
+        }
+
+        buffer
+    }
+
+    fn len(&self) -> usize {
+        SIZE_FIRST + SIZE_SECOND
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        let start = ptr::addr_of!(self.first) as *const u8;
+        unsafe { slice::from_raw_parts(start, self.len()) }
+    }
+
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        let start = ptr::addr_of_mut!(self.first) as *mut u8;
+        unsafe { slice::from_raw_parts_mut(start, self.len()) }
+    }
+}
 
 impl<const SIZE: usize> Serialize for Multihash<SIZE> {
     fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: Serializer,
     {
-        serializer.serialize_bytes(&self.to_bytes())
+        let mut buffer = Buffer::<MAXIMUM_PREFIX_SIZE, SIZE>::new();
+        let bytes_written = self
+            .write(buffer.as_mut_slice())
+            .map_err(|_| ser::Error::custom("Failed to serialize Multihash"))?;
+        serializer.serialize_bytes(&buffer.as_slice()[..bytes_written])
     }
 }
 
@@ -32,7 +88,7 @@ impl<'de, const SIZE: usize> Visitor<'de> for BytesVisitor<SIZE> {
         E: de::Error,
     {
         Multihash::<SIZE>::from_bytes(bytes)
-            .map_err(|err| de::Error::custom(format!("Failed to deserialize Multihash: {}", err)))
+            .map_err(|_| de::Error::custom("Failed to deserialize Multihash"))
     }
 
     // Some Serde data formats interpret a byte stream as a sequence of bytes (e.g. `serde_json`).
@@ -40,12 +96,21 @@ impl<'de, const SIZE: usize> Visitor<'de> for BytesVisitor<SIZE> {
     where
         A: SeqAccess<'de>,
     {
-        let mut bytes = Vec::new();
+        let mut buffer = Buffer::<MAXIMUM_PREFIX_SIZE, SIZE>::new();
+        let bytes = buffer.as_mut_slice();
+
+        // Fill the bytes slices with the given sequence
+        let mut pos = 0;
         while let Some(byte) = seq.next_element()? {
-            bytes.push(byte);
+            bytes[pos] = byte;
+            pos += 1;
+            if pos >= bytes.len() {
+                return Err(de::Error::custom("Failed to deserialize Multihash"));
+            }
         }
-        Multihash::<SIZE>::from_bytes(&bytes)
-            .map_err(|err| de::Error::custom(format!("Failed to deserialize Multihash: {}", err)))
+
+        Multihash::<SIZE>::from_bytes(&bytes[..pos])
+            .map_err(|_| de::Error::custom("Failed to deserialize Multihash"))
     }
 }
 

--- a/src/serde.rs
+++ b/src/serde.rs
@@ -1,6 +1,6 @@
 //! Multihash Serde (de)serialization
 
-use core::{fmt, ptr, slice};
+use core::{fmt, mem, ptr, slice};
 
 use serde::{
     de::{self, SeqAccess, Visitor},
@@ -27,37 +27,20 @@ struct Buffer<const SIZE_FIRST: usize, const SIZE_SECOND: usize> {
 #[allow(unsafe_code)]
 impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
     fn new() -> Self {
-        let buffer = Self {
+        Self {
             first: [0; SIZE_FIRST],
             second: [0; SIZE_SECOND],
-        };
-
-        // Make sure that the struct allocated continuous memory, as we exploit that fact with the
-        // `as_slice` and `as_mut_slice()` methods.
-        let start_first = ptr::addr_of!(buffer.first) as *const u8;
-        let start_second = ptr::addr_of!(buffer.second) as *const u8;
-        unsafe {
-            // This should never happen, hence it's OK to panic.
-            if start_second.offset_from(start_first) != SIZE_FIRST as isize {
-                panic!("alignment of the struct is wrong")
-            }
         }
-
-        buffer
-    }
-
-    fn len(&self) -> usize {
-        SIZE_FIRST + SIZE_SECOND
     }
 
     fn as_slice(&self) -> &[u8] {
         let start = ptr::addr_of!(self.first) as *const u8;
-        unsafe { slice::from_raw_parts(start, self.len()) }
+        unsafe { slice::from_raw_parts(start, mem::size_of::<Self>()) }
     }
 
     fn as_mut_slice(&mut self) -> &mut [u8] {
         let start = ptr::addr_of_mut!(self.first) as *mut u8;
-        unsafe { slice::from_raw_parts_mut(start, self.len()) }
+        unsafe { slice::from_raw_parts_mut(start, mem::size_of::<Self>()) }
     }
 }
 
@@ -236,5 +219,21 @@ mod tests {
                 Token::SeqEnd,
             ],
         );
+    }
+
+    #[test]
+    fn test_buffer_alignment() {
+        const SIZE_FIRST: usize = 11;
+        const SIZE_SECOND: usize = 13;
+        let buffer = Buffer::<SIZE_FIRST, SIZE_SECOND>::new();
+
+        // Make sure that the struct allocated continuous memory, as we exploit that fact with the
+        // `as_slice` and `as_mut_slice()` methods.
+        let start_first = ptr::addr_of!(buffer.first) as *const u8;
+        let start_second = ptr::addr_of!(buffer.second) as *const u8;
+        #[allow(unsafe_code)]
+        unsafe {
+            assert_eq!(start_second.offset_from(start_first), SIZE_FIRST as isize);
+        };
     }
 }

--- a/src/serde.rs
+++ b/src/serde.rs
@@ -1,6 +1,6 @@
 //! Multihash Serde (de)serialization
 
-use core::{fmt, mem, ptr, slice};
+use core::{fmt, mem, slice};
 
 use serde::{
     de::{self, SeqAccess, Visitor},
@@ -34,13 +34,11 @@ impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_
     }
 
     fn as_slice(&self) -> &[u8] {
-        let start = ptr::addr_of!(self.first) as *const u8;
-        unsafe { slice::from_raw_parts(start, mem::size_of::<Self>()) }
+        unsafe { slice::from_raw_parts(self as *const _ as _, mem::size_of::<Self>()) }
     }
 
     fn as_mut_slice(&mut self) -> &mut [u8] {
-        let start = ptr::addr_of_mut!(self.first) as *mut u8;
-        unsafe { slice::from_raw_parts_mut(start, mem::size_of::<Self>()) }
+        unsafe { slice::from_raw_parts_mut(self as *mut _ as _, mem::size_of::<Self>()) }
     }
 }
 
@@ -109,6 +107,8 @@ impl<'de, const SIZE: usize> Deserialize<'de> for Multihash<SIZE> {
 #[cfg(test)]
 mod tests {
     use super::*;
+
+    use std::ptr;
 
     use serde_test::{assert_tokens, Token};
 
@@ -235,5 +235,16 @@ mod tests {
         unsafe {
             assert_eq!(start_second.offset_from(start_first), SIZE_FIRST as isize);
         };
+    }
+
+    #[test]
+    fn test_buffer() {
+        const SIZE_FIRST: usize = 3;
+        const SIZE_SECOND: usize = 8;
+        let mut buffer = Buffer::<SIZE_FIRST, SIZE_SECOND>::new();
+
+        let data: [u8; SIZE_FIRST + SIZE_SECOND] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+        buffer.as_mut_slice().copy_from_slice(&data);
+        assert_eq!(buffer.as_slice(), data);
     }
 }

